### PR TITLE
storage: fix correctness issue of load generators

### LINKED ISSUE #16259: TPC-H load generator source duplicates data upon `environmentd` restart when `TICK INTERVAL` is specified
### What version of Materialize are you using?

v0.33.0-dev (d9bda8a97)

### How did you install Materialize?

Built from source

### What is the issue?

Starting from `bin/environmentd --release --reset` locally, issue the following commands:

```
CREATE DATABASE tpch;
CREATE SOURCE lgtpch FROM LOAD GENERATOR TPCH (SCALE FACTOR 0.1, TICK INTERVAL '0.1s') FOR ALL TABLES WITH (SIZE ='4');
```

At this point, we can observe the number of rows in of one of the relations, `region`, that does not get affected by streaming updates:

```
tpch=> SELECT count(*) FROM region;
 count 
-------
     5
(1 row)
```

Now, restart the system with `bin/environmentd --release`. We can observe the number of rows in `region` again:

```
tpch=> SELECT count(*) FROM region;
 count 
-------
    10
(1 row)
```

These sizes are twice as large as expected. Another restart of `environmentd` will add the rows again, making the TPC-H relations grow one more time.

Note that it is important to specify `TICK INTERVAL` above. If the generator is created without a `TICK INTERVAL` specification, then the problem does not manifest.

Reference to relevant thread in [Slack](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1669239387792779?thread_ts=1669220818.522629&cid=C01CFKM1QRF).

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by petrosagg**
The load generator sources had a correctnes issue where upon restart they would fast foward exactly as many rows as the offset of their resumption frontier.

This only works under the assumption that each offset contains exactly one row which is true for the counter generator but not true for the TPCH generator.

The fix is to simply count the number of transactions when fast fowarding the iterator.

Fixes #16259

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/source/generator.rs [hunk 1]
```diff
@@ -106,8 +106,14 @@ impl SourceConnectionBuilder for LoadGeneratorSourceConnection {
             .by_seed(mz_ore::now::SYSTEM_TIME.clone(), None);
 
         // Skip forward to the requested offset.
+        let mut transactions = rows
+            .by_ref()
+            .filter(|(_, typ, _, _)| matches!(typ, GeneratorMessageType::Finalized));
         for _ in 0..offset.offset {
-            rows.next();
+            assert!(
```
Comment (mjibson) [inline src/storage/src/source/generator.rs:113]: Is it safe to depend on rust not optimizing this away in some build configurations?

**Review by mjibson — APPROVED** [review]

**Review by guswynn — APPROVED** [review]
nice!

File: src/storage/src/source/generator.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/source/generator.rs:113]: This is a great point that I didn't think about! It's probably safe but I'm going to pull the side effects out of the assert to be certain

Comment (petrosagg) [reply src/storage/src/source/generator.rs]: Fixed!

Comment (bkirwi) [reply src/storage/src/source/generator.rs]: (FWIW, `assert!` is [guaranteed to be run in every possible build config](https://doc.rust-lang.org/std/macro.assert.html#uses). Doesn't hurt to move this out, though!)

Comment (petrosagg) [reply src/storage/src/source/generator.rs]: That's good to know! thanks for the link

**Pull request merged**